### ISSUE TITLE
groupby // apply  breaks on a larger Dataframe

### ISSUE THREAD

**Issue by Arengard** [pola-rs/polars#1870]
#### Are you using Python or Rust?
Python 3.10
#### What version of polars are you using?
0.10.23
#### What operating system are you using polars on?
Ubuntu 20.04/Windows 10 
#### Describe your bug.
grouping on a large dataframe and applying a  UDF breaks !!!

```
import polars as pl
am = pl.read_csv('reddit_vm.csv').with_row_count()

# some arbitrary udf
def test_apply(df):
    abc = str(df.title[0])
    cde = str(df.id[0])
    amd = str(df.score[0]*2)
    return df.with_column(pl.when(pl.col('title')!='').then(amd +'_____'+ cde +'_____'+ abc).otherwise('ne').alias('test_apply'))

done = am.groupby('row_nr').apply(test_apply).sort('row_nr')

# works like Charm :) 
done.head()

# but if the Frame is larger
nd = pl.concat([am]*5_000)
nd.shape

# it breaks 
done = nd .groupby('row_nr').apply(test_apply).sort('row_nr')


```

[reddit_vm.csv](https://github.com/pola-rs/polars/files/7589581/reddit_vm.csv)

**Comment by ritchie46** [comment]
What do you mean it breaks?

**Comment by Arengard** [comment]
kernel dies

**Comment by ritchie46** [comment]
> kernel dies

Then you are out of memory. 

Applying over subdataframes is also not recommended. It is slow and memory hungry.

What do you exactly want to do? I think we can write it in the expression API.

**Comment by ghuls** [comment]
@ritchie46 It looks like there is also some real bugs and not just an out of memory error.

```python
In [7]: def test(multiplyer):
   ...:     nd = pl.concat([am] * multiplyer)
   ...:     print(multiplyer, nd.shape)
   ...:     nd.groupby('row_nr').apply(test_apply).sort('row_nr').head()
   ...:     del nd
   ...: 

In [8]: for i in [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]:
   ...:     test(i)
   ...: 
20 (31940, 9)
21 (33537, 9)
Segmentation fault (core dumped)

```

Also when using the expression API, some underflow/overflow occurs when it tries to allocate memory (with size of dataframe only twice of `am`:

```python
In [1]: import polars as pl
   ...: am = pl.read_csv('reddit_vm.csv').with_row_count()
   ...: 
   ...: # some arbitrary udf
   ...: def test_apply(df):
   ...:     abc = str(df.title[0])
   ...:     cde = str(df.id[0])
   ...:     amd = str(df.score[0]*2)
   ...:     return df.with_column(pl.when(pl.col('title')!='').then(amd +'___
   ...: __'+ cde +'_____'+ abc).otherwise('ne').alias('test_apply'))
   ...: 
   ...: done = am.groupby('row_nr').apply(test_apply).sort('row_nr')
   ...: 
   ...: # works like Charm :) 
   ...: done.head()


In [7]: done.head()
Out[7]: 
shape: (5, 10)
┌────────┬──────────────┬───────┬────────┬─────┬──────────────┬──────┬──────────────┬──────────────┐
│ row_nr ┆ title        ┆ score ┆ id     ┆ ... ┆ created      ┆ body ┆ timestamp    ┆ test_apply   │
│ ---    ┆ ---          ┆ ---   ┆ ---    ┆     ┆ ---          ┆ ---  ┆ ---          ┆ ---          │
│ u32    ┆ str          ┆ i64   ┆ str    ┆     ┆ f64          ┆ str  ┆ datetime     ┆ str          │
╞════════╪══════════════╪═══════╪════════╪═════╪══════════════╪══════╪══════════════╪══════════════╡
│ 0      ┆ Health       ┆ 7     ┆ lt74vw ┆ ... ┆ 1.614400425e ┆      ┆ 2021-02-27   ┆ 14_____lt74v │
│        ┆ Canada       ┆       ┆        ┆     ┆ 9            ┆      ┆ 06:33:45     ┆ w_____Health │
│        ┆ approves     ┆       ┆        ┆     ┆              ┆      ┆              ┆ Canada ...   │
│        ┆ AstraZene... ┆       ┆        ┆     ┆              ┆      ┆              ┆              │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1      ┆ COVID-19 in  ┆ 2     ┆ lsh0ij ┆ ... ┆ 1.614316267e ┆      ┆ 2021-02-26   ┆ 4_____lsh0ij │
│        ┆ Canada: 'Vac ┆       ┆        ┆     ┆ 9            ┆      ┆ 07:11:07     ┆ _____COVID-1 │
│        ┆ cination...  ┆       ┆        ┆     ┆              ┆      ┆              ┆ 9 in Can...  │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2      ┆ Coronavirus  ┆ 6     ┆ lohlle ┆ ... ┆ 1.613886608e ┆      ┆ 2021-02-21   ┆ 12_____lohll │
│        ┆ variants     ┆       ┆        ┆     ┆ 9            ┆      ┆ 07:50:08     ┆ e_____Corona │
│        ┆ could fuel   ┆       ┆        ┆     ┆              ┆      ┆              ┆ virus va...  │
│        ┆ ...          ┆       ┆        ┆     ┆              ┆      ┆              ┆              │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 3      ┆ Canadian     ┆ 1     ┆ lnptv8 ┆ ... ┆ 1.613795713e ┆      ┆ 2021-02-20   ┆ 2_____lnptv8 │
│        ┆ government   ┆       ┆        ┆     ┆ 9            ┆      ┆ 06:35:13     ┆ _____Canadia │
│        ┆ to extend    ┆       ┆        ┆     ┆              ┆      ┆              ┆ n govern...  │
│        ┆ CO...        ┆       ┆        ┆     ┆              ┆      ┆              ┆              │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 4      ┆ Canada:      ┆ 6     ┆ lkslm6 ┆ ... ┆ 1.613468188e ┆      ┆ 2021-02-16   ┆ 12_____lkslm │
│        ┆ Pfizer is    ┆       ┆        ┆     ┆ 9            ┆      ┆ 11:36:28     ┆ 6_____Canada │
│        ┆ 'extremely   ┆       ┆        ┆     ┆              ┆      ┆              ┆ : Pfizer...  │
│        ┆ com...       ┆       ┆        ┆     ┆              ┆      ┆              ┆              │
└────────┴──────────────┴───────┴────────┴─────┴──────────────┴──────┴──────────────┴──────────────┘

In [9]: nd = pl.concat([am]*2)

In [29]: am.groupby("row_nr", maintain_order=True).agg(
    ...:     [
    ...:         pl.when(
    ...:             pl.col("title").first() != ""
    ...:         ).then(
    ...:             (pl.col("score") * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title")
    ...:         ).otherwise(
    ...:             "ne"
    ...:         ).alias("test_apply")
    ...:     ]
    ...: )
Out[29]: 
shape: (1597, 2)
┌────────┬─────────────────────────────────────┐
│ row_nr ┆ test_apply                          │
│ ---    ┆ ---                                 │
│ u32    ┆ list [str]                          │
╞════════╪═════════════════════════════════════╡
│ 0      ┆ ["14_____lt74vw_____Health Canad... │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1      ┆ ["4_____lsh0ij_____COVID-19 in C... │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2      ┆ ["12_____lohlle_____Coronavirus ... │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 3      ┆ ["2_____lnptv8_____Canadian gove... │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ ...    ┆ ...                                 │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1592   ┆ ["2_____elrrt5l_____Comment"]       │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1593   ┆ ["2_____elrqfk3_____Comment"]       │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1594   ┆ ["2_____elrph9m_____Comment"]       │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1595   ┆ ["2_____elroxqn_____Comment"]       │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1596   ┆ ["2_____elro7rb_____Comment"]       │
└────────┴─────────────────────────────────────┘

# Forgot pl.col("score").first() here:
In [30]: nd.groupby("row_nr", maintain_order=True).agg(
    ...:     [
    ...:         pl.when(
    ...:             pl.col("title").first() != ""
    ...:         ).then(
    ...:             (pl.col("score") * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title")
    ...:         ).otherwise(
    ...:             "ne"
    ...:         ).alias("test_apply")
    ...:     ]
    ...: )
memory allocation of 6872316418862805990 bytes failed
Aborted (core dumped)


# With pl.col("score").first():

In [3]: am.groupby("row_nr", maintain_order=True).agg(
   ...:     [
   ...:         pl.when(
   ...:             pl.col("title").first() != ""
   ...:         ).then(
   ...:             (pl.col("score").first() * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title").first()
   ...:         ).otherwise(
   ...:             "ne"
   ...:         ).alias("test_apply")
   ...:     ]
   ...: )
Out[3]: 
shape: (1597, 2)
┌────────┬─────────────────────────────────────┐
│ row_nr ┆ test_apply                          │
│ ---    ┆ ---                                 │
│ u32    ┆ list [str]                          │
╞════════╪═════════════════════════════════════╡
│ 0      ┆ ["14_____lt74vw_____Health Canad... │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1      ┆ ["4_____lsh0ij_____COVID-19 in C... │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2      ┆ ["12_____lohlle_____Coronavirus ... │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 3      ┆ ["2_____lnptv8_____Canadian gove... │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ ...    ┆ ...                                 │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1592   ┆ ["2_____elrrt5l_____Comment"]       │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1593   ┆ ["2_____elrqfk3_____Comment"]       │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1594   ┆ ["2_____elrph9m_____Comment"]       │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1595   ┆ ["2_____elroxqn_____Comment"]       │
├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 1596   ┆ ["2_____elro7rb_____Comment"]       │
└────────┴─────────────────────────────────────┘


# on the nd dataframe, I get most of the time a segfault, bu als omanaged to get a PanicException once:

In [4]: nd.groupby("row_nr", maintain_order=True).agg(
   ...:     [
   ...:         pl.when(
   ...:             pl.col("title").first() != ""
   ...:         ).then(
   ...:             (pl.col("score").first() * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title").first()
   ...:         ).otherwise(
   ...:             "ne"
   ...:         ).alias("test_apply")
   ...:     ]
   ...: )
Segmentation fault (core dumped)


In [10]: nd.groupby("row_nr", maintain_order=True).agg(
    ...:     [
    ...:         pl.when(
    ...:             pl.col("title").first() != ""
    ...:         ).then(
    ...:             (pl.col("score").first() * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title").first()
    ...:         ).otherwise(
    ...:             "ne"
    ...:         ).alias("test_apply")
    ...:     ]
    ...: )
thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /github/home/.cargo/git/checkouts/arrow2-8a2ad61d97265680/70562fa/src/array/utf8/mutable.rs:640:46
---------------------------------------------------------------------------
PanicException                            Traceback (most recent call last)
<ipython-input-10-203c2c5b95e0> in <module>
----> 1 nd.groupby("row_nr", maintain_order=True).agg(
      2     [
      3         pl.when(
      4             pl.col("title").first() != ""
      5         ).then(

~/software/anaconda3/envs/polars/lib/python3.8/site-packages/polars/internals/frame.py in agg(self, column_to_agg)
   3449             elif isinstance(column_to_agg[0], pli.Expr):
   3450                 return (
-> 3451                     wrap_df(self._df)
   3452                     .lazy()
   3453                     .groupby(self.by, maintain_order=self.maintain_order)

~/software/anaconda3/envs/polars/lib/python3.8/site-packages/polars/internals/lazy_frame.py in collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, string_cache, no_optimization)
    291             string_cache,
    292         )
--> 293         return pli.wrap_df(ldf.collect())
    294 
    295     def fetch(

PanicException: called `Option::unwrap()` on a `None` value

```

**Comment by ritchie46** [comment]
> @ritchie46 It looks like there is also some real bugs and not just an out of memory error.
> 
> ```python
> In [7]: def test(multiplyer):
>    ...:     nd = pl.concat([am] * multiplyer)
>    ...:     print(multiplyer, nd.shape)
>    ...:     nd.groupby('row_nr').apply(test_apply).sort('row_nr').head()
>    ...:     del nd
>    ...: 
> 
> In [8]: for i in [20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30]:
>    ...:     test(i)
>    ...: 
> 20 (31940, 9)
> 21 (33537, 9)
> Segmentation fault (core dumped)
> ```
> 
> Also when using the expression API, some underflow/overflow occurs when it tries to allocate memory (with size of dataframe only twice of `am`:
> 
> ```python
> In [1]: import polars as pl
>    ...: am = pl.read_csv('reddit_vm.csv').with_row_count()
>    ...: 
>    ...: # some arbitrary udf
>    ...: def test_apply(df):
>    ...:     abc = str(df.title[0])
>    ...:     cde = str(df.id[0])
>    ...:     amd = str(df.score[0]*2)
>    ...:     return df.with_column(pl.when(pl.col('title')!='').then(amd +'___
>    ...: __'+ cde +'_____'+ abc).otherwise('ne').alias('test_apply'))
>    ...: 
>    ...: done = am.groupby('row_nr').apply(test_apply).sort('row_nr')
>    ...: 
>    ...: # works like Charm :) 
>    ...: done.head()
> 
> 
> In [7]: done.head()
> Out[7]: 
> shape: (5, 10)
> ┌────────┬──────────────┬───────┬────────┬─────┬──────────────┬──────┬──────────────┬──────────────┐
> │ row_nr ┆ title        ┆ score ┆ id     ┆ ... ┆ created      ┆ body ┆ timestamp    ┆ test_apply   │
> │ ---    ┆ ---          ┆ ---   ┆ ---    ┆     ┆ ---          ┆ ---  ┆ ---          ┆ ---          │
> │ u32    ┆ str          ┆ i64   ┆ str    ┆     ┆ f64          ┆ str  ┆ datetime     ┆ str          │
> ╞════════╪══════════════╪═══════╪════════╪═════╪══════════════╪══════╪══════════════╪══════════════╡
> │ 0      ┆ Health       ┆ 7     ┆ lt74vw ┆ ... ┆ 1.614400425e ┆      ┆ 2021-02-27   ┆ 14_____lt74v │
> │        ┆ Canada       ┆       ┆        ┆     ┆ 9            ┆      ┆ 06:33:45     ┆ w_____Health │
> │        ┆ approves     ┆       ┆        ┆     ┆              ┆      ┆              ┆ Canada ...   │
> │        ┆ AstraZene... ┆       ┆        ┆     ┆              ┆      ┆              ┆              │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1      ┆ COVID-19 in  ┆ 2     ┆ lsh0ij ┆ ... ┆ 1.614316267e ┆      ┆ 2021-02-26   ┆ 4_____lsh0ij │
> │        ┆ Canada: 'Vac ┆       ┆        ┆     ┆ 9            ┆      ┆ 07:11:07     ┆ _____COVID-1 │
> │        ┆ cination...  ┆       ┆        ┆     ┆              ┆      ┆              ┆ 9 in Can...  │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 2      ┆ Coronavirus  ┆ 6     ┆ lohlle ┆ ... ┆ 1.613886608e ┆      ┆ 2021-02-21   ┆ 12_____lohll │
> │        ┆ variants     ┆       ┆        ┆     ┆ 9            ┆      ┆ 07:50:08     ┆ e_____Corona │
> │        ┆ could fuel   ┆       ┆        ┆     ┆              ┆      ┆              ┆ virus va...  │
> │        ┆ ...          ┆       ┆        ┆     ┆              ┆      ┆              ┆              │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 3      ┆ Canadian     ┆ 1     ┆ lnptv8 ┆ ... ┆ 1.613795713e ┆      ┆ 2021-02-20   ┆ 2_____lnptv8 │
> │        ┆ government   ┆       ┆        ┆     ┆ 9            ┆      ┆ 06:35:13     ┆ _____Canadia │
> │        ┆ to extend    ┆       ┆        ┆     ┆              ┆      ┆              ┆ n govern...  │
> │        ┆ CO...        ┆       ┆        ┆     ┆              ┆      ┆              ┆              │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌┼╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 4      ┆ Canada:      ┆ 6     ┆ lkslm6 ┆ ... ┆ 1.613468188e ┆      ┆ 2021-02-16   ┆ 12_____lkslm │
> │        ┆ Pfizer is    ┆       ┆        ┆     ┆ 9            ┆      ┆ 11:36:28     ┆ 6_____Canada │
> │        ┆ 'extremely   ┆       ┆        ┆     ┆              ┆      ┆              ┆ : Pfizer...  │
> │        ┆ com...       ┆       ┆        ┆     ┆              ┆      ┆              ┆              │
> └────────┴──────────────┴───────┴────────┴─────┴──────────────┴──────┴──────────────┴──────────────┘
> 
> In [9]: nd = pl.concat([am]*2)
> 
> In [29]: am.groupby("row_nr", maintain_order=True).agg(
>     ...:     [
>     ...:         pl.when(
>     ...:             pl.col("title").first() != ""
>     ...:         ).then(
>     ...:             (pl.col("score") * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title")
>     ...:         ).otherwise(
>     ...:             "ne"
>     ...:         ).alias("test_apply")
>     ...:     ]
>     ...: )
> Out[29]: 
> shape: (1597, 2)
> ┌────────┬─────────────────────────────────────┐
> │ row_nr ┆ test_apply                          │
> │ ---    ┆ ---                                 │
> │ u32    ┆ list [str]                          │
> ╞════════╪═════════════════════════════════════╡
> │ 0      ┆ ["14_____lt74vw_____Health Canad... │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1      ┆ ["4_____lsh0ij_____COVID-19 in C... │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 2      ┆ ["12_____lohlle_____Coronavirus ... │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 3      ┆ ["2_____lnptv8_____Canadian gove... │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ ...    ┆ ...                                 │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1592   ┆ ["2_____elrrt5l_____Comment"]       │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1593   ┆ ["2_____elrqfk3_____Comment"]       │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1594   ┆ ["2_____elrph9m_____Comment"]       │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1595   ┆ ["2_____elroxqn_____Comment"]       │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1596   ┆ ["2_____elro7rb_____Comment"]       │
> └────────┴─────────────────────────────────────┘
> 
> # Forgot pl.col("score").first() here:
> In [30]: nd.groupby("row_nr", maintain_order=True).agg(
>     ...:     [
>     ...:         pl.when(
>     ...:             pl.col("title").first() != ""
>     ...:         ).then(
>     ...:             (pl.col("score") * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title")
>     ...:         ).otherwise(
>     ...:             "ne"
>     ...:         ).alias("test_apply")
>     ...:     ]
>     ...: )
> memory allocation of 6872316418862805990 bytes failed
> Aborted (core dumped)
> 
> 
> # With pl.col("score").first():
> 
> In [3]: am.groupby("row_nr", maintain_order=True).agg(
>    ...:     [
>    ...:         pl.when(
>    ...:             pl.col("title").first() != ""
>    ...:         ).then(
>    ...:             (pl.col("score").first() * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title").first()
>    ...:         ).otherwise(
>    ...:             "ne"
>    ...:         ).alias("test_apply")
>    ...:     ]
>    ...: )
> Out[3]: 
> shape: (1597, 2)
> ┌────────┬─────────────────────────────────────┐
> │ row_nr ┆ test_apply                          │
> │ ---    ┆ ---                                 │
> │ u32    ┆ list [str]                          │
> ╞════════╪═════════════════════════════════════╡
> │ 0      ┆ ["14_____lt74vw_____Health Canad... │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1      ┆ ["4_____lsh0ij_____COVID-19 in C... │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 2      ┆ ["12_____lohlle_____Coronavirus ... │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 3      ┆ ["2_____lnptv8_____Canadian gove... │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ ...    ┆ ...                                 │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1592   ┆ ["2_____elrrt5l_____Comment"]       │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1593   ┆ ["2_____elrqfk3_____Comment"]       │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1594   ┆ ["2_____elrph9m_____Comment"]       │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1595   ┆ ["2_____elroxqn_____Comment"]       │
> ├╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
> │ 1596   ┆ ["2_____elro7rb_____Comment"]       │
> └────────┴─────────────────────────────────────┘
> 
> 
> # on the nd dataframe, I get most of the time a segfault, bu als omanaged to get a PanicException once:
> 
> In [4]: nd.groupby("row_nr", maintain_order=True).agg(
>    ...:     [
>    ...:         pl.when(
>    ...:             pl.col("title").first() != ""
>    ...:         ).then(
>    ...:             (pl.col("score").first() * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title").first()
>    ...:         ).otherwise(
>    ...:             "ne"
>    ...:         ).alias("test_apply")
>    ...:     ]
>    ...: )
> Segmentation fault (core dumped)
> 
> 
> In [10]: nd.groupby("row_nr", maintain_order=True).agg(
>     ...:     [
>     ...:         pl.when(
>     ...:             pl.col("title").first() != ""
>     ...:         ).then(
>     ...:             (pl.col("score").first() * 2).cast(pl.Utf8) + pl.lit("_____") + pl.col("id").first() + pl.lit("_____") + pl.col("title").first()
>     ...:         ).otherwise(
>     ...:             "ne"
>     ...:         ).alias("test_apply")
>     ...:     ]
>     ...: )
> thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', /github/home/.cargo/git/checkouts/arrow2-8a2ad61d97265680/70562fa/src/array/utf8/mutable.rs:640:46
> ---------------------------------------------------------------------------
> PanicException                            Traceback (most recent call last)
> <ipython-input-10-203c2c5b95e0> in <module>
> ----> 1 nd.groupby("row_nr", maintain_order=True).agg(
>       2     [
>       3         pl.when(
>       4             pl.col("title").first() != ""
>       5         ).then(
> 
> ~/software/anaconda3/envs/polars/lib/python3.8/site-packages/polars/internals/frame.py in agg(self, column_to_agg)
>    3449             elif isinstance(column_to_agg[0], pli.Expr):
>    3450                 return (
> -> 3451                     wrap_df(self._df)
>    3452                     .lazy()
>    3453                     .groupby(self.by, maintain_order=self.maintain_order)
> 
> ~/software/anaconda3/envs/polars/lib/python3.8/site-packages/polars/internals/lazy_frame.py in collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, string_cache, no_optimization)
>     291             string_cache,
>     292         )
> --> 293         return pli.wrap_df(ldf.collect())
>     294 
>     295     def fetch(
> 
> PanicException: called `Option::unwrap()` on a `None` value
> ```

Thanks, I will take a look

**Comment by ritchie46** [comment]
Fixed in #1872
